hw/formal/formal_conn.sh

```diff
@@ -0,0 +1,108 @@
+#!/bin/bash
+
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+# Script to run connectivity test using Jasper Gold
+#
+# Usage: To run connectivity test on system level modules, type
+#   formal_conn -f top_earlgrey.csv
+#
+# More options:
+# -top: which top_level module to run
+# -p:  provide core file path
+# -gui: run with GUI
+# -cov: run coverage
+# -f: path to the csv file
+# -t: choose which formal tool to use, current only support jaspergold
+#
+# Example:
+#   formal_conn -gui -cov -f top_earlgrey.csv
+#
+
+export TOP="top_earlgrey"
+gui=0
+tool="jg"
+export COV=0
+export CSV_PATH
+
+while [ "$1" != "" ]; do
+  case "$1" in
+    "-top")
+      shift
+      TOP=$1
+      ;;
+    "-p")
+      shift
+      CORE_PATH=$1
+      ;;
+    "-gui")
+      gui=1
+      echo "using GUI mode"
+      ;;
+    "-cov")
+      COV=1
+      echo "enable coverage"
+      ;;
+    "-t")
+      shift
+      tool=$1
+      echo "running in tool: $tool"
+      ;;
+    "-f")
+      shift
+      CSV_PATH=../../../conn_csvs/$1
+      echo "csv path: $CSV_PATH"
+      ;;
+  esac
+  shift
+done
+
+echo "-------------------------------------------------------------------------"
+echo "-- Generate file list using FuseSoC"
+echo "-------------------------------------------------------------------------"
+echo ""
+echo "Top: $TOP"
+echo ""
+
+\rm -Rf build jgproject
+# we just run the setup for the default target in order to generate the filelist
+if [ "${CORE_PATH}" == "" ]; then
+  CORE_PATH=systems:top_earlgrey
+fi
+echo "core_file path: lowrisc:${CORE_PATH}"
+
+fusesoc --cores-root ../.. run\
+        --tool=icarus --target=default\
+        --flag=fileset_top\
+        --setup "lowrisc:${CORE_PATH}"
+
+echo "-------------------------------------------------------------------------"
+echo "-- Run JasperGold"
+echo "-------------------------------------------------------------------------"
+
+cd build/*${TOP}*/default-icarus
+
+if [ "${tool}" == "jg" ]; then
+  if [ "${gui}" == "1" ]; then
+    jg ../../../tools/jaspergold/conn.tcl \
```

Comment:
It might be clearer to use `readlink -f` to get an absolute path to the top of the repo and then use that here. That way, you can also support different build directories in the future if you want.

Reply:
I see, thanks Rupert. I added a `readlink -f` at the beginning of the script. Thank you for pointing that out.

Reply:
Hmm, that's not quite what I meant: I meant that you could use `readlink -f` to get the absolute path to the top of the repository. You can use that for `TCL_SCRIPT_PATH` (as you've done), but also for all the other things that start with `../../..` in the file.